### ISSUE TITLE
Unable able to switch to WebView on real android device or on a simulator

### ISSUE THREAD

**Issue by abhijeetkanade** [appium/appium#5089]
Hi,

I trying to test hybrid app, so trying a sample test on a selendroid-test-app. There are two issues with WebView on a android.

Tried following code to get context names:

for (String contextName : contextNames) {

System.out.println(contextName); //prints out something like NATIVE_APP \n WEBVIEW_1

//driver.context(contextName);

}
1. On real device, it only returns NATIVE_APP view, so can't even try to switch to web view on a real device.
2. on a simulator, we do get a web view. e.g. for selendroid app, above code returns

NATIVE_APP

WEBVIEW_io.selendroid.testapp

however while trying to switch to webview getting error

driver.context("WEBVIEW_io.selendroid.testapp");

org.openqa.selenium.WebDriverException: An unknown server-side error occurred while processing the command. (Original error: Did not get session redirect from Chromedriver) (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 5.97 seconds
Build info: version: '2.45.0', revision: '5017cb8e7ca8e37638dc3091b2440b90a1d8686f', time: '2015-02-27 09:10:26'
System info: host: 'abhijeet-PC', ip: '192.168.1.4', os.name: 'Windows 7', os.arch: 'amd64', os.version: '6.1', java.version: '1.7.0_21'
Session ID: 389c9143-9a3f-456f-8e23-251b515579b1
Driver info: io.appium.java_client.android.AndroidDriver
Capabilities [{newCommandTimeout=120, platformVersion=4.1, app=C:\Users\abhijeet\Google Drive\Selenium\Android Projects\MavenApp2\app\selendroid-test-app-0.15.0.apk, platform=LINUX, databaseEnabled=false, javascriptEnabled=true, platformName=Android, deviceName=Android device, browserName=Android, webStorageEnabled=false, networkConnectionEnabled=true, desired={newCommandTimeout=120, app=C:\Users\abhijeet\Google Drive\Selenium\Android Projects\MavenApp2\app\selendroid-test-app-0.15.0.apk, deviceName=Android device, platformName=Android}, locationContextEnabled=false, warnings={}, takesScreenshot=true}]
    at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
    at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
    at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
    at java.lang.reflect.Constructor.newInstance(Unknown Source)
    at org.openqa.selenium.remote.ErrorHandler.createThrowable(ErrorHandler.java:204)
    at org.openqa.selenium.remote.ErrorHandler.throwIfResponseFailed(ErrorHandler.java:156)
    at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:599)
    at io.appium.java_client.AppiumDriver.execute(AppiumDriver.java:180)
    at io.appium.java_client.AppiumDriver.context(AppiumDriver.java:525)
    at tests.AppiumSample.selendroidTest1(AppiumSample.java:53)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
    at java.lang.reflect.Method.invoke(Unknown Source)
    at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
    at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
    at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
    at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
    at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
    at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
    at org.testng.TestRunner.privateRun(TestRunner.java:767)
    at org.testng.TestRunner.run(TestRunner.java:617)
    at org.testng.SuiteRunner.runTest(SuiteRunner.java:334)
    at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:329)
    at org.testng.SuiteRunner.privateRun(SuiteRunner.java:291)
    at org.testng.SuiteRunner.run(SuiteRunner.java:240)
    at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
    at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
    at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
    at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
    at org.testng.TestNG.run(TestNG.java:1057)
    at org.testng.remote.RemoteTestNG.run(RemoteTestNG.java:111)
    at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:204)
    at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:175)

request you to please look into matter.

**Comment by amaanmemon** [comment]
Hey, there is no need to switch to webview for real android device, it automatically open google chrome once you provide the proper desired capabilities
or else
driver.context("WEB_VIEW")

**Comment by abhijeetkanade** [comment]
driver.context("WEB_VIEW") is throwing following error
io.appium.java_client.NoSuchContextException: No such context found. (WARNING: The server did not provide any stacktrace information)
Command duration or timeout: 195 milliseconds

**Comment by amaanmemon** [comment]
driver = new AppiumDriver(new URL("http://127.0.0.1:4723/wd/hub"), capabilities);

Set<String> contextNames = driver.getContextHandles();
for (String contextName : contextNames) {
    System.out.println(contextNames); //prints out something like NATIVE_APP \n WEBVIEW_1
}
driver.context(contextNames.toArray()[1]); // set context to WEBVIEW_1

//do some web testing
String myText = driver.findElement(By.cssSelector(".green_button")).click();

driver.context("NATIVE_APP");

// do more native testing if we want

driver.quit();

**Comment by abhijeetkanade** [comment]
capa.setCapability(MobileCapabilityType.PLATFORM_NAME, MobilePlatform.ANDROID);		
		capa.setCapability(MobileCapabilityType.DEVICE_NAME, "Android device");		
		capa.setCapability(MobileCapabilityType.NEW_COMMAND_TIMEOUT, 120);
		capa.setCapability(MobileCapabilityType.BROWSER_NAME,"");
		capa.setCapability(MobileCapabilityType.APP_PACKAGE, packageName);
		capa.setCapability(MobileCapabilityType.APP_ACTIVITY, activityName);
		

		AndroidDriver driver= new AndroidDriver(new URL("http://0.0.0.0:4723/wd/hub"), capa);


I am defining all the capabilities, but instead of AppiumDriver I am using AndroidDriver. and using following code to print all the contexts:
  Set<String> contextNames = driver.getContextHandles();
	  for (String contextName : contextNames) {
	      System.out.println(contextName); //prints out something like NATIVE_APP \n WEBVIEW_1
	  }
for real device it only prints NATIVE_APP and for simulator it prints NATIVE_AP and WEBVIEW_io.selendroid.testapp

**Comment by vijay0850** [comment]
I'm also facing same issue on real device it showing NATIVE_APP only but here driver.context(contextNames.toArray()[1]);

When the array is set to 1 it throwing ArrayInputBound Exception.

**Comment by Jonahss** [comment]
So Android real devices aren't showing their webview contexts?

**Comment by vijay0850** [comment]
Yeap, It displaying Native App only, In my case I'm using API level less than 19.  I set the selendroid compatibility also.

I have one more device having API level 19 for this setting autoWebView compatibility is enough?

**Comment by Jonahss** [comment]
The low API level might be the issue. This is hard for me to test because I don't have an android phone on that API level.
It works fine in the emulators?

**Comment by abhijeetkanade** [comment]
No, I tested issue with API level 21 on a real device, it did not work. It only returns Native_App on a real device. On a simulator, it does return webview context, however while switching to Webview context throws an error which I have mentioned in my first post.

**Comment by vijay0850** [comment]
@Jonahss, In API level 19 also I'm not able to switch to webview, please take look into this below screen shot  

![image](https://cloud.githubusercontent.com/assets/10095173/7604952/9d6c9da2-f967-11e4-9d86-e6bcbfce159b.png)

Code is:- ( For API 19 We no need to set compatibility to selenroid rite)

![image](https://cloud.githubusercontent.com/assets/10095173/7605005/3173ff7c-f968-11e4-9bba-9eb6fc27f26a.png)

**Comment by vijay0850** [comment]
Any update on this issue, I'm still facing issue

**Comment by bchand1748** [comment]
I has a similar issue few months back where I was not able to switch the context eventhough the code seem to be right . The issue was fixed once I put thread.sleep 
Thread.sleep(15000)
 Set<String> contextNames = driver.getContextHandles();
	     for (String contextName : contextNames) {
	       System.out.println(contextName);
	       if (contextName.contains("WEBVIEW")){
	         driver.context(contextName);
	     	Thread.sleep(15000);
Also, I was not able to switch context using Appium. Therefore, I decided to use Selendroid while using the Appium GUI. In the GUI change the automation name to selendriod and on your code set the capabilities to capabilities.setCapability("automationName","selendroid");

**Comment by abhijeetkanade** [comment]
@bchand1748 did you check it recently? I do have Thread.sleep(10000) in my code, but it still returns NATIVE_APP only. I wasn't lucky with capa.setCapability(MobileCapabilityType.AUTOMATION_NAME,"Selendroid");I am getting error with it, Could you please paste complete code and steps to test with Selendroid using Appium.

**Comment by vijay0850** [comment]
Hi, @abhijeetkanade as per @bchand1748 suggestion it worked for me in one app.

Here is the code I used to switch from native to webview.

package com.test.appium;

import io.appium.java_client.AppiumDriver;
import io.appium.java_client.android.AndroidDriver;

import java.io.File;
import java.net.URL;
import java.util.Set;
import java.util.concurrent.TimeUnit;

import org.openqa.selenium.By;
import org.openqa.selenium.remote.DesiredCapabilities;

public class HdfcWebviewtest {
	
	private static AppiumDriver driver;


	public static void main(String arg[]) throws Exception {
		
		// set up appium		
		 DesiredCapabilities capabilities = new DesiredCapabilities();
		 File appDir = new File("D://Appium//AppTest//");
		 File app = new File(appDir, "com.elance.mobile.apk");
		 capabilities.setCapability("appium-version", "1.3.7");
		 capabilities.setCapability("deviceName","0123456789ABCDEF");
		 capabilities.setCapability("platformVersion", "4.2.2");
		 capabilities.setCapability("platformName","Android");
		 capabilities.setCapability("autoWebview","true");
		 capabilities.setCapability("automationName","selendroid");
		 capabilities.setCapability("newCommandTimeout", "180");
		 capabilities.setCapability("app", "com.elance.mobile");
		 capabilities.setCapability("app", app.getAbsolutePath());
		 capabilities.setCapability("appActivity", ".ElancePhoneGapActivity");
		 driver = new AndroidDriver(new URL("http://127.0.0.1:4723/wd/hub"),capabilities);
		 		    
		 Thread.sleep(15000);
		 Set<String> contextNames = driver.getContextHandles();
		 for (String contextName : contextNames) {
		 System.out.println(contextName);
		 if (contextName.contains("WEBVIEW")){
		 driver.context(contextName);
		 System.out.println(contextName);
		 Thread.sleep(1500);	
		 
		 driver.findElement(By.xpath("//html/body/div/section/form/div[1]/label/input")).sendKeys("vijay0850");
	
	
		 }

		 }

	}
	
}

**Comment by abhijeetkanade** [comment]
@vijay0850 Hi Vijay, I still have no luck here. I tried with capabilities.setCapability("automationName","selendroid"); (I even tried as Selendroid)in my code but it throwing null pointer exception. 
FAILED CONFIGURATION: @AfterMethod afterMethod
java.lang.NullPointerException
	at tests.AppiumSample.afterMethod(AppiumSample.java:115)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)

Are you sure capabilities.setCapability("automationName","selendroid") works with API 21?
I am using API level 21 and java client 2.2.0

**Comment by TianleZhang** [comment]
In my test environment. The Webview context appear time to time. very unstable and I have no idea about it. Really hope this issue can be fix soon.

**Comment by michalmikolajczyk** [comment]
+1, same results, please see also 
https://discuss.appium.io/t/cant-switch-context-to-webview/4539
https://discuss.appium.io/t/webview-context-is-not-found-for-ios-app/4535

**Comment by imurchie** [comment]
Can you try, on the command line, to run the command that should list the webviews, which Appium uses to get them? And then post the results (there will be a lot, not all relevant) here?

```shell
adb shell "cat /proc/net/unix"
```

(If you have multiple devices attached, you will need to use the `-s` switch, like `adb -s 4d00bfca4c9c5131 shell "cat /proc/net/unix"` where `4d00bfca4c9c5131` is the id of the device.)

**Comment by michalmikolajczyk** [comment]
@imurchie is there a similar one for iOS that I could post to help to solve this issue?

**Comment by imurchie** [comment]
A similar issue? Or a similar command? This issue is about Android only. Mixing them makes it more difficult to work on.

**Comment by michalmikolajczyk** [comment]
Well, the issue that I experience on iOS, seems effectively equal to what is described above.
I was directed here from the appium forums, the links are provided above.

Do you think I should make another task?

**Comment by sebv** [comment]
Moving to bug, keep discussing...keep it Android please.

**Comment by kchuykov** [comment]
I am not able to see webview on this app. The webview did show up only once and I was not able to switch to it. 
I tried adb shell "cat /proc/net/unix" command and here is the output: https://www.dropbox.com/s/rn382fb9qd2mb40/adb_command.txt?dl=0

app file:  https://www.dropbox.com/s/law3xry54szdq69/app.apk?dl=0
on the first screen click deposit funds to get to the web view page.

**Comment by dmlemeshko** [comment]
Hi guys, I'm also facing the same issue on Nexus 5 device (Android 5.1.1) and Huawei 3C (Android 4.2.2) for hybrid app. Works on emulator. Are there any short term plans for the fix?

**Comment by jackywyz** [comment]
I have the same issue on Samsung phone (Android 4.4.2)

**Comment by ShankarGargXebia** [comment]
Set contextNames = driver.getContextHandles();
	Iterator iter = contextNames.iterator();
	while (iter.hasNext()) {
	    System.out.println(iter.next());
	}
	Thread.sleep(2500);

	driver.context("WEBVIEW_com.apptivateme.next.ct");
	Thread.sleep(2500);

	final WebElement inputField = driver.findElement(By.name("email"));
	inputField.sendKeys("abc");

worked for me with Appium 1.4.8 and Java

Earlier i was using driver.findElement(By.name("email")).sendKeys("abc") which was not working for me. Not sure why it is like this. but for the moment it worked for me.

**Comment by carolinasim** [comment]
Hello everyone!
There are new solutions for this problem?

I'm facing something similar. I'm working with an hybrid app and the Appium only returns the NATIVE_APP context. I can see the app in chrome inspector so I presume that the setWebContentsDebuggingEnabled is set to true.

Can you suggest something?

**Comment by vinodbhardwaj1** [comment]
Hi All,
is this mandatory to set "setWebContentsDebuggingEnabled=true" for hybrid app automation on android Lollipop also? since on production /play store apk, can't keep it true (due to security concern as per dev team). 

If we create script on staging env with 'true' how will we run the scripts on prod app.

Please suggest !

**Comment by githubramakrishnan** [comment]
hi friends,

just Rebuilt the apk with set view enabled (true).
that's all, now the application will return with both Native and Web View.

**Comment by vinodbhardwaj1** [comment]
Ya Thanks,

I did the same and working fine on staging env app.

On Tue, Dec 8, 2015 at 10:31 AM, githubramakrishnan <
notifications@github.com> wrote:

> hi friends,
>
> just Rebuilt the apk with set view enabled (true).
> that's all, now the application will return with both Native and Web View.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/appium/appium/issues/5089#issuecomment-162763409>.
>



-- 

Thanks n Regards

*------------------------*Vinod Kumar Bhardwaj
*9899946472*

**Comment by RavindraSHirandagi** [comment]
Hi  vinodbhardwaj1 or githubramakrishnan,
please tell me where to set this one set view enabled (true).

**Comment by githubramakrishnan** [comment]
Hi Ravi,

just place the line --> set view enabled (true) in the app and rebuilt the app
thats all.

**Comment by RavindraSHirandagi** [comment]
Hello @githubramakrishnan i am not getting where and in which file i have to modify, please tell me clearly the file name and where to modify..

Thanks

**Comment by githubramakrishnan** [comment]
hi Ravi ,

Enable the webview by using the line **set view enabled (true)**  in  **.js**  file.

**Comment by cattm** [comment]
Hi there All. Apologies if this is out of order - having read the entire thread, this issue sounds exactly the one I am having.  I am using the same App (it works fine on 4.3 phone and emulator). But when I move to 4.4 and upwards it exhibits the behaviours discussed above. If I have understood correctly the feeling is that the problem is the "app" itself and I need to grab the source and modify that. Have I understood correctly?

**Comment by Antony11** [comment]
Hi All , 

Has anyone else found a workaround to getting WebView context on a real Android device ?

**Comment by fdx321** [comment]
I faced the same problem,  appium is ****, macaca is better http://macacajs.github.io/macaca/

**Comment by KateOgar** [comment]
Are there some updates here? still actual on appium 1.5.2

**Comment by kskrishnan** [comment]
I am also still stuck with the same issue. I can't see WEBVIEW context on both Emulator and the Real Device. In my case, the WEBVIEW is actually a new browser window for Twitter authentication. This browser window is spawned by the Gigya APIs and I am not sure how to control this using setWebContentsDebuggingEnabled=TRUE or by set view enabled (true) in .js file. Can someone be explicit as to which .js file is that?

**Comment by Bakalaka** [comment]
Did you have installed facebook app ?

**Comment by navin578** [comment]
I am also having this issue 'switching between windows ' since long time only in devices i.e., android and IOS.
I am unable to replicate the issue ,I even tried it using switching between tabs which worked fine in all browsers but when I tried to run the same code in devices error occurred saying "NoSuchWindowException: A request to switch to a different window could not be satisfied because the window could not be found."

I tried in many ways but failed , I tried to add the capability of allowing pop-ups and in some situations both URLS opens in same window( I mean first window url and secound window url in same window).

Can anyone please tell me what exactly causing this problem.

**Comment by jrgoulet** [comment]
Experiencing the same issue:

Android 5.1, Appium 4.0.0, Selenium 1.5.2

getContextHandles() only returns NATIVE_APP for Chrome Browser (49).

**Comment by Sudheendran** [comment]
Did this issue got resolved? Pls help us.

**Comment by jrgoulet** [comment]
I actually figured out that I only needed to switch to NATIVE_APP once towards the beginning of an automated test, and I was able to selected all the elements I needed to between the OS and different applications from within this context.

I mistakenly thought that there were additional contexts that I was not seeing.

**Comment by rajeesat** [comment]
I am facing the similar issue when run automation tests on Android real device(Nexus 5) OS version 6.0.1.Any updates on this?Please help me if anyone have an idea. Thanks in advance

**Comment by kundanps** [comment]
Hi 
i am also getting the same problem , i am working on a hybrid app where i have to switch from WEBVIEW to NATIVE, perform some action on native and then again switch to WEBVIEW, i am able to perform action on NATIVE but could not switch back to WEBVIEW.
Here is the code which i am using to switch to WEBVIEW
Set<String> contextNames = browser.getContextHandles();
        for (String contextName : contextNames) {
            if (contextName.contains("WEBVIEW")) {
                log("Switched to Web View Context.");
                browser.context(contextName);
            }
        }

Any help will be highly appreciated.

**Comment by s-sridevi15** [comment]
I am also seeing same issue . I cannot see webView context on appium . We are using reactNative and by default reactNative enables setWebContentDebuggingEnalbe for debug.apk files with below condition, which should show webview for app-debug.apk file(but not app-release.apk). 

if (ReactBuildConfig.DEBUG && Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
      WebView.setWebContentsDebuggingEnabled(true);
    }
@githubramakrishnan, we have webView.js file. could you please provide more details with actual command how I can enalble web view enabled(true) in js file. The line you gave "set view enabled (true)" to enable on js, is this the right command? It does not work if we add just like that with spaces. Would you mind sharing the right command and is it webView.js which I need to enable?

Does this enabling works for reactNative builds on real devices?

**Comment by Akshay64** [comment]
I am also facing the same issue. I am getting WEBVIEW context when i run test on Android Emulator but when I run the test on real android devices, it returns on NATIVE context.

Anybody? plz help

**Comment by dam1** [comment]
If you are using Crosswalk you need to use the [patched chromedriver webdriver](https://github.com/rokkknrollla/appium-android-xwalk-driver-patch) and use the context "CHROMIUM"

**Comment by Akshay64** [comment]
@dam1 will you pleas elaborate what is crosswalk

**Comment by zrleo** [comment]
hi everyone i am facing the same issue when i run my test case on my phone （Android 4.4.2），There is only NATIVE_APP context!

Is anybody resolved this issue！！！plz help me！！！Thanks！！！

**Comment by jlipps** [comment]
Unfortunately this issue has turned into a catch-all for a variety of problems. I'm going to close it. If you are still having a problem, please update your android SDKs and Appium versions to the latest, and recreate issues with all the appropriate repro information.
